core/src/parser/expression.rs

```diff
@@ -0,0 +1,103 @@
+pub use super::parser::*;
+pub use super::operator::*;
+pub use super::ast::*;
+pub use super::literal::*;
+pub use crate::lexer::*;
+
+pub fn ParseExpression(parser: Parser) -> Expr {
+    ParseExpressionWithPrecedence(parser, -1)
+}
+
+pub fn ParseExpressionWithPrecedence(parser:Parser, precedence: i32) -> Expr {
+    let Token::Operator(operator) = parser.peek().unwrap();
+
+    let previous = Expr::Unary(ParseUnaryOrPrimaryExpression(parser));
+    
+    //TODO: precedence
+    if Operator::is_binary(operator) {
+        return ParseBinaryExpression(parser, previous);
+    } else {
+        return previous
+    }
+}
+
+pub fn ParseUnaryOrPrimaryExpression(parser: Parser) -> UnaryExpr {
+    let Token::Operator(operator) = parser.peek().unwrap();
+
+    if Operator::is_unary(operator) {
+        return UnaryExpr::UnaryOp {
+            op: ParseUnaryOperator(parser),
+            expr: Box::new(ParseUnaryOrPrimaryExpression(parser))
+        }
+    } else {
+        return ParsePrimaryExpression(parser)
+    };
+}
+
+pub fn ParseBinaryExpression(parser:Parser, previous: Expr) -> Expr {
+    Expr::Binary{
+        lhs: Box::new(previous),
+        op: ParseBinaryOperator(parser),
+        //TODO: precedence
+        rhs: Box::new(ParseExpressionWithPrecedence(parser , 0))
+    }
+}
+
+// pub fn ParseUnaryExpression(parser:Parser) -> UnaryExpr {
+//     UnaryExpr::UnaryOp {
+//         op: ParseUnaryOperator(parser),
+//         expr: Box::new(ParseUnaryExpression(parser))
+//     }
+// }
+
+pub fn ParsePrimaryExpression(parser:Parser) -> UnaryExpr {
+    let current = parser.peek().unwrap();
+
+    //TODO: ParseConditional, ParseSelector
+    let primary:PrimaryExpr = match current {
+        Token::BoolLit(current) => ParseBooleanLiteral(parser),
+        Token::IntLit(current) => ParseIntegerLiteral(parser),
+        Token::FloatLit(current) => ParseFloatLiteral(parser),
+        Token::CharLit(current) => ParseCharLiteral(parser),
+        Token::StrLit(current) => ParseStringLiteral(parser),
+        Token::Ident(current) => 
+            if parser.is_tok_curr(Token::Punc(Punc::LeftParen), parser.peek_offset(1)) {
+                ParseFunctionCall(parser)
+            } else {
+                ParseVariableAccess(parser)
+            }
+        Token::Punc(Punc::LeftParen) => ParseGrouping(parser)
+    };
+    UnaryExpr::Primary(primary)
+}
+
+pub fn ParseFunctionCall(parser:Parser) -> PrimaryExpr {
+    let Some(Token::Ident(name)) = parser.eat_type(Token::Ident(String::new()));
+    
+    parser.eat_type(Token::Punc(Punc::LeftParen));
```

Comment:
We could probably make repetition (https://whistle.js.org/docs/specification/grammar#notation) more abstract